### PR TITLE
Expose discrete type info

### PR DESCRIPTION
**Pull request by pszmel**
### Details:
 - Expose DiscreteTypeInfo class

### Tickets:
 - 57594

### CODE REVIEW

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (jiwaszki) [inline ngraph/python/src/pyngraph/discrete_type_info.cpp:22]: Looks OK, but this can be moved to some common space with helper functions.

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (jiwaszki) [inline ngraph/python/src/pyngraph/discrete_type_info.cpp:60]: Same thing applies here - common space with helpers.

**Review by jiwaszki — COMMENTED** [review]
Looks good, waiting for tests to approve 🚀

**Review by jiwaszki — APPROVED** [review]
LGTM

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (postrational) [inline ngraph/python/src/pyngraph/discrete_type_info.cpp:24]: We should avoid using `new` here.

File: ngraph/python/src/pyngraph/common.cpp
Comment (postrational) [inline ngraph/python/src/pyngraph/common.cpp:6]: I think it should be possible to avoid having this function.

**Review by postrational — CHANGES_REQUESTED** [review]

**Review by tomdol — CHANGES_REQUESTED** [review]
I suggest not allowing the python API users to create random instances of this class. If it's needed in the python API at all, I'd rather just add the "read-only" bindings allowing to get the DiscreteTypeInfo for OV objects. I don't think the constructor should be there at all.

The current version will leak the memory allocated for the char* because it's not freed anywhere.

@ilyachur @ilya-lavrenov what do you guys think?

**Comment by ilyachur** [comment]
> I suggest not allowing the python API users to create random instances of this class. If it's needed in the python API at all, I'd rather just add the "read-only" bindings allowing to get the DiscreteTypeInfo for OV objects. I don't think the constructor should be there at all.
> 
> The current version will leak the memory allocated for the char* because it's not freed anywhere.
> 
> @ilyachur @ilya-lavrenov what do you guys think?

@tomdol The write access to `DiscreteTypeInfo` is needed only if user want to create a custom operation from the Python API, but if I understand correctly this case is not supported.
But if we are talking about aligning of C++ and Python API we are using this structure in `get_type_info()` method.
So I propose to think how it should look in Python API, and maybe we can provide the python equivalent of `DiscreteTypeInfo` or something like that.

File: ngraph/python/src/pyngraph/node.cpp
Comment (jiwaszki) [inline ngraph/python/src/pyngraph/node.cpp:41]: Can it be moved after all of the operators to keep "style" of registering?

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (jiwaszki) [inline ngraph/python/src/pyngraph/discrete_type_info.cpp:64]: I would like to see `__repr__` return something in style of: `<DiscreteTypeInfo: [OpName] v[OpVersion] Parent([ParentName] v[ParentVersion])>`, please keep spaces between version etc

Also if Parent is `nullptr` I would consider not giving the Parent's info at all - but this is not required feature.

File: ngraph/python/tests/test_ngraph/test_core.py
Comment (jiwaszki) [inline ngraph/python/tests/test_ngraph/test_core.py:242]: Why is this `K` a capital?

**Review by jiwaszki — CHANGES_REQUESTED** [review]

File: ngraph/python/src/pyngraph/node.cpp
Comment (tomdol) [inline ngraph/python/src/pyngraph/node.cpp:28]: Not sure but is the last trailing comma needed here?

**Review by tomdol — APPROVED** [review]

File: ngraph/python/src/pyngraph/node.cpp
Comment (jiwaszki) [reply ngraph/python/src/pyngraph/node.cpp:28]: It follows the convention from pybind's documentation, as in examples this is the way to inform that no arguments are used in underlying function. I can be kept in my opinion.

File: ngraph/python/src/pyngraph/node.cpp
Comment (pszmel) [reply ngraph/python/src/pyngraph/node.cpp]: Moved.

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (pszmel) [reply ngraph/python/src/pyngraph/discrete_type_info.cpp]: Updated.

File: ngraph/python/tests/test_ngraph/test_core.py
Comment (pszmel) [reply ngraph/python/tests/test_ngraph/test_core.py]: Typo. Fixed.

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (jiwaszki) [inline ngraph/python/src/pyngraph/discrete_type_info.cpp:68]: Remove `else` and just return name + version in case the parent is a nullptr so function always returns something.

**Review by jiwaszki — CHANGES_REQUESTED** [review]
One small change and it is ready to merge.

**Review by jiwaszki — APPROVED** [review]

File: ngraph/python/src/pyngraph/discrete_type_info.cpp
Comment (postrational) [inline ngraph/python/src/pyngraph/discrete_type_info.cpp:29]: Looks like you're duplicating code from `DiscreteTypeInfo::operator<=` here. We should be able to reuse these `operator` functions here, rather than recreating their logic again.

**Review by postrational — CHANGES_REQUESTED** [review]

**Review by postrational — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

File: runtime/bindings/python/src/compatibility/pyngraph/discrete_type_info.cpp
Comment (postrational) [inline runtime/bindings/python/src/compatibility/pyngraph/discrete_type_info.cpp:40]: Can we use PyBind to bind these properties directly like this?
```suggestion
    discrete_type_info.def_property_readonly("name", &ngraph::DiscreteTypeInfo::name);
```

**Review by postrational — CHANGES_REQUESTED** [review]

File: runtime/bindings/python/src/compatibility/pyngraph/discrete_type_info.cpp
Comment (postrational) [reply runtime/bindings/python/src/compatibility/pyngraph/discrete_type_info.cpp:40]: You may need to use `def_readonly` in this case.
https://pybind11.readthedocs.io/en/latest/classes.html#instance-and-static-fields

Comment (pszmel) [reply runtime/bindings/python/src/compatibility/pyngraph/discrete_type_info.cpp]: Fixed.

**Review by postrational — APPROVED** [review]

**Review by akuporos — APPROVED** [review]

**Pull request merged**
